### PR TITLE
Add Local and Unstructured resource builder attributes and handle <rsrsc> / <name> pairs when --local is set

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:
```release-note
NONE
```

Related downstream BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1390139

Adds an error message if a non-filename (or stdin) resource is specified with the `--local` flag in any of the `kube set ...` sub-commands.

cc @fabianofranz @kubernetes/sig-cli-misc

### CODE REVIEW

**Comment by juanvallejo** [comment]
@fabianofranz wondering if this is good to go? Thanks!

**Comment by fabianofranz** [comment]
@juanvallejo This works for me, but here goes an idea. I didn't check the feasibility so it may not even work. 

Anyways: would it make sense to make "local" an attribute of the `Builder` like every other (e.g. `builder.Local(local).SelectorParam(...)...`, then in the `DisabledClientForMapping`, instead of returning `nil, nil` in `ClientForMapping` you'd return a new `RESTClient` that encapsulates a logic that, anytime it would try to make an API call, it would return an error? 

Main concept of the idea would be to rely on the actual client and, in any tentative of an API call, throw the error. Would fix it for any new commands we add in future, an avoid the need for so many conditionals.

**Comment by fabianofranz** [comment]
> Would fix it for any new commands we add in future

Also I remember fixing a few days ago a panic related to that `nil, nil` and some obscure corner case.  :)

**Comment by juanvallejo** [comment]
@fabianofranz 
> Anyways: would it make sense to make "local" an attribute of the Builder like every other (e.g. builder.Local(local).SelectorParam(...)..., then in the DisabledClientForMapping, instead of returning nil, nil in ClientForMapping you'd return a new RESTClient that encapsulates a logic that, anytime it would try to make an API call, it would return an error?

I think this would certainly improve the way we would handle local command invocations.
Will give it a try as part of this PR

**Comment by juanvallejo** [comment]
@fabianofranz https://github.com/kubernetes/kubernetes/pull/48763/commits/ecfa3945baff14c738beb78c9bbd462359b02421 adds a new `Local` and `Unstructured` attribute, removing the need for `NewUnstructuredBuilder` factory method. PTAL

File: pkg/kubectl/resource/mapper.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ type DisabledClientForMapping struct {
 }
 
 func (f DisabledClientForMapping) ClientForMapping(mapping *meta.RESTMapping) (RESTClient, error) {
-	return nil, nil
+	return nil, fmt.Errorf("attempt to make api call using a disabled client")
```
Comment (juanvallejo) [inline pkg/kubectl/resource/mapper.go]: @fabianofranz not sure if this is the right approach, or if it's better to just make and return a fake RESTClient

**Comment by juanvallejo** [comment]
@fabianofranz Not sure how to fix bazel test. Getting the error:
```
missing strict dependencies:
W0803 19:38:31.906] 	pkg/kubectl/resource/builder.go: import of k8s.io/apimachinery/pkg/apis/meta/v1/unstructured, which is not a direct dependency
```

Tried adding this line to `pkg/kubectl/resource/BUILD` but that does not seem to fix it:
`"//vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured:go_default_library"`

**Comment by juanvallejo** [comment]
cc @lavalamp

**Comment by juanvallejo** [comment]
@lavalamp or @smarterclayton friendly ping.
Wondering if you could approve this PR

**Comment by juanvallejo** [comment]
cc @kubernetes/sig-cli-maintainers

**Comment by juanvallejo** [comment]
cc @smarterclayton

**Comment by cblecker** [comment]
```
W0831 18:16:30.714] # k8s.io/kubernetes/pkg/kubectl/cmd/set
W0831 18:16:30.715] pkg/kubectl/cmd/set/set_env.go:234: too many arguments in call to f.NewBuilder
W0831 18:16:30.715] 	have (bool)
W0831 18:16:30.715] 	want ()
W0831 18:16:30.716] pkg/kubectl/cmd/set/set_env.go:296: too many arguments in call to f.NewBuilder
W0831 18:16:30.716] 	have (bool)
W0831 18:16:30.716] 	want ()
```
This isn't building. Removing LGTM to allow for fix.
/lgtm cancel

**Comment by juanvallejo** [comment]
@fabianofranz had to rebase to bring in the `set_env` cmd and update the resource builder there. Wondering if you can take one more look at this? Thanks!

**Comment by juanvallejo** [comment]
@cblecker or @fabianofranz wondering if you could /lgtm this once more? Thanks!

File: pkg/kubectl/resource/builder.go [hunk 2]
```diff
@@ -155,6 +167,35 @@ func (b *Builder) FilenameParam(enforceNamespace bool, filenameOptions *Filename
...
+
+	b.isLocal = true
+	b.mapper.ClientMapper = DisabledClientForMapping{ClientMapper: ClientMapperFunc(mapperFunc)}
+	return b
+}
+
+// Unstructured updates the builder's ClientMapper, RESTMapper,
+// ObjectTyper, and codec for working with unstructured api objects
+func (b *Builder) Unstructured(mapperFunc ClientMapperFunc, mapper meta.RESTMapper, typer runtime.ObjectTyper) *Builder {
+	if b.isLocal {
```
Comment (smarterclayton) [inline pkg/kubectl/resource/builder.go]: This is not the right place to check this.  Local and unstructured have nothing to do with each other.  Local commands should absolutely be able to function without needing to have compiled in objects.

Comment (juanvallejo) [reply pkg/kubectl/resource/builder.go]: @smarterclayton Thanks, will make a patch to address this

**Pull request merged**